game/src/core/components/PlayerComponent.java

```diff
@@ -60,21 +62,22 @@ public Optional<Consumer<Entity>> registerFunction(int key, Consumer<Entity> fun
     /**
      * Remove the registered function on the given key.
      *
-     * @param key Value of the key.
+     * @param key The integer value of the key.
+     * @see com.badlogic.gdx.Gdx#input
      */
-    public void removeFunction(int key) {
+    public void removeFunction(final int key) {
         functions.remove(key);
     }
 
     /**
-     * Will check each entry in the function map, and if the key is just pressed, the function will
-     * be executed.
+     * This method will check each entry in the function map, and if the key is just pressed, the
+     * function will be executed.
      */
     public void execute() {
-        functions.forEach((k, f) -> execute(k, f));
+        functions.forEach(this::execute);
     }
 
-    private void execute(Integer key, Consumer<Entity> function) {
+    private void execute(final int key, final Consumer<Entity> function) {
         if (Gdx.input.isKeyPressed(key)) function.accept(entity);
```

Comment:
@AMatutat Ich bin noch nicht so richtig glücklich mit diesem Design.

(a) Ich muss irgendwo gültige Tastaturcodes herkriegen und hier dezentral speichern und mit einem Callback verknüpfen. (Implizite Abhängigkeit zu libGDX, verstreut über den Code/mehrere potentielle Instanzen von PlayerComponent?)
(b) Ich muss jedesmal alle registrierten Tastaturcodes durchgehen _und_ libGDX fragen, ob einer davon gedrückt wurde. (Aufwand, Abhängigkeit zu libGDX).
(c) Beim Aufruf gebe ich den registrierten Callbacks die Entität mit. Dann stellt sich die Frage, warum ich die Funktionen hier in der Component speichere (wenn sie quasi auch auf anderen Entitäten arbeiten könnten) oder warum ich sie nicht bereits beim registrieren "personalisiere" und dann beim Aufruf die Parameter-Übergabe spare?

Wie wäre es, wenn man die konkreten libGDX-Tastaturcodes in einer Enum kapselt und hier im Framework nur noch die Enum-Konstanten nutzt? Außerdem könnte das PlayerSystem die Abfrage der gedrückten Tasten machen und die entsprechenden Enum-Konstanten dann hier mit reinreichen. Dann habe ich nur noch an zwei Stellen die Abhängigkeit zu libGDX (Enum, PlayerSystem).

Noch besser wäre es, man würde aus der PlayerComponent heraus eine Reihe von "Commands" erzeugen (Objekte mit einer `execute`-Methode und verknüpft mit einer Entität (dieser hier)) und diese beim PlayerSystem für eine Enum-Konstante (Button, Tastatureingabe) registrieren. Dann könnte das PlayerSystem bei libGDX erfragen, was grad gedrückt wurde und in der internen Map nachschauen, welche Commands für exakt diese Eingabe registriert wurden und für diese dann `execute` aufrufen. => [Command-Pattern](https://www.hsbi.de/elearning/data/FH-Bielefeld/lm_data/lm_1359639/pattern/command.html#command-objektorientierte-antwort-auf-callback-funktionen) plus [Observer-Pattern](https://www.hsbi.de/elearning/data/FH-Bielefeld/lm_data/lm_1359639/pattern/observer.html#observer-pattern-verallgemeinert)

Reply:
#760